### DISCUSSION TITLE
Is HEAD supposed to return content-length ?

### DISCUSSION THREAD

**Discussion by Ex4l** [vercel/next.js#37684] [Help]
Hello, we are using NextJS in production for our website and we noticed that each HEAD request is sending a content length in the HEAD request response.

It's a bit problematic for us as we need to cache those pages through our CDN.

CURL example on localhost : 

````
curl -v -X HEAD http://localhost:3000/account
Warning: Setting custom HTTP method to HEAD with -X/--request may not work the
Warning: way you want. Consider using -I/--head instead.
*   Trying 127.0.0.1:3000...
* Connected to localhost (127.0.0.1) port 3000 (#0)
> HEAD /account HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.79.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Server: nginx/1.21.6
< Date: Tue, 14 Jun 2022 13:05:47 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 2047
< Connection: keep-alive
< Cache-Control: no-store, must-revalidate
< ETag: "7ff-LsLbg2R68RG2rxVPIby9T0L46Mk"
< Vary: Accept-Encoding
<
* transfer closed with 2047 bytes remaining to read
* Closing connection 0
curl: (18) transfer closed with 2047 bytes remaining to read
````

We have checked RFC 9110 HTTP Semantics, which allowes this content-length but it seems to be better not having this header.

RFC 9110 HTTP Semantics:
> A server MAY send a Content-Length header field in a response to a HEAD request (Section 9.3.2); a server MUST NOT send Content-Length in such a response unless its field value equals the decimal number of octets that would have been sent in the content of a response if the same request had used the GET method.

Is that behavior normal or not ? Could we fix it or not ?

**Comment by adamoleeo** [comment]
NextJS is correct, you're using curl wrong :)

https://curl.se/docs/manpage.html#-X

> This option only changes the actual word used in the HTTP request, it does not alter the way curl behaves. For example if you want to make a proper HEAD request, using -X HEAD does not suffice. You need to use the [-I, --head](https://curl.se/docs/manpage.html#-I) option.

So curl is effectively treating the response like it would a GET response and waiting for the body. Use `curl -v --head http://localhost:3000/account` instead.
